### PR TITLE
Create CRUD for member without DB

### LINKED ISSUE #20: Create a new CRUD resource called discord
Use NestJS CLI to create a new CRUD resource called `discord` with the generate command `resource` [docs](https://docs.nestjs.com/cli/usages#nest-generate)

This is a subtask (https://github.com/EddieJaoudeCommunity/api/issues/12)

### LINKED ISSUE #21: Create CRUD for member without DB
This is a subtask (https://github.com/EddieJaoudeCommunity/api/issues/12)
```ts
interface Member {
   username: string;
   bio: {
      description: string;
      twitter: string;
      linkedin: string;
      github: string;
   };
   updatedOn: Date;
   createdOn: Date;
}
```

### PR DESCRIPTION
**Pull request by khattakdev**
This is how Discord object looks like
```ts
export class CreateDiscordDto{
  id: number
   username: string;
   bio: {
      description: string;
      twitter: string;
      linkedin: string;
      github: string;
   };
   updatedOn: Date;
   createdOn: Date;
}
```
I've added `id` temporary, to track down objects for the update, single search, and deletion. Once we have a database, we can remove it, if we don't need it but for now, it's something needed.

I've tested the API(s) as well, and they are working perfectly fine.

**Get Single Record**
getOne

**Update Record**
patch

**Insert Record**
post

**Delete Record**
delete

**Get All Records**
getAll

Closes #21

### CODE REVIEW

File: src/discord/discord.service.ts [hunk 1]
```diff
@@ -1,26 +1,73 @@
...
+      updatedDiscord.bio.twitter = twitter;
+    }
+    if (linkedin) {
+      updatedDiscord.bio.linkedin = linkedin;
+    }
+    if (github) {
+      updatedDiscord.bio.github = github;
+    }
+
+    return `User updated successfully!`;
```
Comment (eddiejaoude) [inline src/discord/discord.service.ts:55]: The return string on line `61` uses single quotes, it is better to be consistent with that if there are no variables in the string here

File: src/discord/discord.service.ts [hunk 2]
```diff
@@ -1,26 +1,73 @@
...
 
   remove(id: number) {
-    return `This action removes a #${id} discord`;
+    const userIndex = this.discord.findIndex((user) => user.id === id);
+    this.discord.splice(userIndex);
+    return 'User deleted successfully!';
+  }
+
+  private findDiscord(id: number): ReadDiscordDto {
+    const userIndex = this.discord.findIndex((user) => user.id === id);
```
Comment (eddiejaoude) [inline src/discord/discord.service.ts:65]: I think a `find` here would be cleaner, where it does not match the id, so it is excluded

File: src/discord/dto/read-discord.dto.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { PartialType } from '@nestjs/mapped-types';
+import { CreateDiscordDto } from './create-discord.dto';
+
+export class ReadDiscordDto extends PartialType(CreateDiscordDto) {
+  id: number;
```
Comment (eddiejaoude) [inline src/discord/dto/read-discord.dto.ts:5]: is `id` needed if it is created from `CreateDiscordDto`?

File: src/discord/discord.service.ts [hunk 4]
```diff
@@ -1,26 +1,73 @@
...
   }
 
   findAll() {
-    return `This action returns all discord`;
+    return [...this.discord];
   }
 
   findOne(id: number) {
-    return `This action returns a #${id} discord`;
+    const discordUser = this.findDiscord(id);
```
Comment (eddiejaoude) [inline src/discord/discord.service.ts:26]: I think this private function call could be replaced with a javascript `find`

File: src/discord/dto/read-discord.dto.ts (hunk 3 above)
Comment (khattakdev) [reply src/discord/dto/read-discord.dto.ts:5]: This line no `5` is making it necessary, `id` exist with that line too but it is optional ( `id?: number` ).

File: src/discord/discord.service.ts (hunk 4 above)
Comment (eddiejaoude) [reply src/discord/discord.service.ts:26]: In the future, the DB will be doing it, so I think it might be optimisation too soon

File: src/discord/dto/read-discord.dto.ts (hunk 3 above)
Comment (eddiejaoude) [reply src/discord/dto/read-discord.dto.ts:5]: I do not understand, where is it optional? It looks like it is required here...? https://github.com/EddieJaoudeCommunity/api/pull/23/files#diff-cdd95909b05df4daa83d67707b55752498545cda37904ddf279494e56aa4bedeR2

File: src/discord/discord.service.ts (hunk 4 above)
Comment (khattakdev) [reply src/discord/discord.service.ts:26]: > I think this private function call could be replaced with a javascript `find`

It's also doing error handling 
```ts
if (!discordUser) {
      throw new NotFoundException('User Not Found');
    }
```
Which I'll have to repeat on multiple locations.  How about I keep the function and replace `findIndex` with `find`? (Although it won't improve performance even when we have large data)

File: src/discord/dto/read-discord.dto.ts (hunk 3 above)
Comment (khattakdev) [reply src/discord/dto/read-discord.dto.ts:5]: I mean to say when we use `PartialType`, it will copy the values and make all of them optional (excluding nested).

Here, we have these properties in `create-discord.dto.ts`
```ts
  id: number;
  username: string;
  bio: {
    description?: string;
    twitter?: string;
    linkedin?: string;
    github?: string;
  };
  updatedOn: Date;
  createdOn: Date;
```

The top-level properties are necessary but in `read-discord.dto.ts`
If I just add
```ts
export class ReadDiscordDto extends PartialType(CreateDiscordDto) {}
```

This will results in `ReadDiscordDto` having the following properties
```ts
  id?: number;
  username?: string;
  bio?: {
    description?: string;
    twitter?: string;
    linkedin?: string;
    github?: string;
  };
  updatedOn?: Date;
  createdOn?: Date;
```

But since `id` is a primary key, which is required in every **API**, so we need to make it necessary. To do that, we can redefine that specific property in the `ReadDiscordDto`
```ts
export class ReadDiscordDto extends PartialType(CreateDiscordDto) {
  id: number;
}
```

Now this will result in
```ts
  id: number;
  username?: string;
  bio?: {
    description?: string;
    twitter?: string;
    linkedin?: string;
    github?: string;
  };
  updatedOn?: Date;
  createdOn?: Date;
```

---
In short, whatever property we define in that class will be compulsory. `PartialType` copy the properties and make it optional, if we need to make any property compulsory, we have to redefine it.

File: src/discord/discord.service.ts (hunk 4 above)
Comment (eddiejaoude) [reply src/discord/discord.service.ts]: the `find` is temporary, because it will be handled by the DB. We can leave this function for now if you like

File: src/discord/discord.service.ts [hunk 5]
```diff
@@ -1,26 +1,71 @@
...
 
   remove(id: number) {
-    return `This action removes a #${id} discord`;
+    const userIndex = this.discord.findIndex((user) => user.id === id);
+    this.DiscordNotFound(userIndex);
+    this.discord.splice(userIndex);
+    return 'User deleted successfully!';
+  }
+
+  private DiscordNotFound(discord: ReadDiscordDto | number) {
```
Comment (eddiejaoude) [inline src/discord/discord.service.ts:66]: This naming does not match the other conventions. I do not think a private method is needed for this, plus it causes type mixing

File: src/discord/discord.service.ts [hunk 6]
```diff
@@ -1,26 +1,71 @@
...
+    if (github) {
+      updatedDiscord.bio.github = github;
+    }
+
+    return 'User updated successfully!';
   }
 
   remove(id: number) {
-    return `This action removes a #${id} discord`;
+    const userIndex = this.discord.findIndex((user) => user.id === id);
```
Comment (eddiejaoude) [inline src/discord/discord.service.ts:60]: `find` would work better here, but still better to use es6 style

It will eventually be replaced by the DB

File: src/discord/discord.service.ts (hunk 5 above)
Comment (khattakdev) [reply src/discord/discord.service.ts:66]: What do you prefer over here? directly use `NotFoundException` instead of making a function for it?

File: src/discord/dto/read-discord.dto.ts (hunk 3 above)
Comment (eddiejaoude) [reply src/discord/dto/read-discord.dto.ts:5]: Ah I see, thank you for clarifying 👍

File: src/discord/discord.service.ts (hunk 6 above)
Comment (khattakdev) [reply src/discord/discord.service.ts:60]: > `find` would work better here.
> 
> It will eventually be replaced by the DB, but still better to use es6 style

I'm using splice in the next line. I am not sure how can we remove an item if we use `find`. How about we use `filter` instead of `findIndex` or `find` and update the whole record.

Comment (khattakdev) [reply src/discord/discord.service.ts:60]: > > `find` would work better here.
> > It will eventually be replaced by the DB, but still better to use es6 style
> 
> I'm using `splice` in the next line. I am not sure how can we remove an item if we use `find`. How about we use `filter` instead of `findIndex` or `find` and update the whole record.

**Note:** `this.discord` is read-only at the moment, we will have to tweak that if we use `filter`

File: src/discord/discord.service.ts (hunk 5 above)
Comment (eddiejaoude) [reply src/discord/discord.service.ts:66]: yes for now please 👍  - so each area can use the object or the number, mixing types is not good

File: src/discord/discord.service.ts (hunk 6 above)
Comment (eddiejaoude) [reply src/discord/discord.service.ts:60]: sorry I meant `filter` not `find`, but yes `filter` out the one we no longer required would be cleaner

**Review by eddiejaoude — APPROVED** [review]
Looks good 👍
